streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -679,41 +711,20 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
   }
 
   /**
-   * Unwraps [[Exit]] values that also signify end-of-stream by failing with `None`.
-   *
-   * For `Exit[E, O]` values that do not signal end-of-stream, prefer:
-   * {{{
-   * stream.mapM(ZIO.done(_))
-   * }}}
+   * Terminates the stream when encountering the first `Left`.
    */
-  def collectWhileSuccess[E1 >: E, O1](implicit ev: O <:< Exit[Option[E1], O1]): ZStream[R, E1, O1] =
-    ZStream {
-      for {
-        upstream <- self.process.mapM(BufferedPull.make(_))
-        done     <- Ref.make(false).toManaged_
-        pull = done.get.flatMap {
-                 if (_) Pull.end
-                 else
-                   upstream.pullElement
-                     .foldM(
-                       {
-                         case None    => done.set(true) *> Pull.end
-                         case Some(e) => Pull.fail(e)
-                       },
-                       os =>
-                         ZIO
-                           .done(ev(os))
-                           .foldM(
-                             {
-                               case None    => done.set(true) *> Pull.end
-                               case Some(e) => Pull.fail(e)
-                             },
-                             Pull.emit(_)
-                           )
-                     )
-               }
-      } yield pull
-    }
+  final def collectWhileRight[L1, O1](implicit ev: O <:< Either[L1, O1]): ZStream[R, E, O1] = {
+    val _ = ev
+    self.asInstanceOf[ZStream[R, E, Either[L1, O1]]].collectWhile { case Right(a) => a }
+  }
+
+  /**
+   * Terminates the stream when encountering the first [[Exit.Success]].
```

Comment:
Doesn't it terminate the stream when encountering the first `Failure`?

Reply:
Oops, yes.